### PR TITLE
Implement `CupertinoListSection` and `CupertinoListTile`

### LINKED ISSUE #19228: Create a CupertinoTableRow and CupertinoTable
The table should offer a grouping function with group label.
The row should offer a basic and a subtitled mode.

Checklist:

- [ ] From https://github.com/flutter/flutter/issues/24604, tables with more content than the page scrolls. Tables with less content than the page creates fake rows with the size of the last item and fills the table with it. That table still overscrolls with more of the fake rows.
- [ ] From https://github.com/flutter/flutter/issues/24605, make sure table can take rows of different height.
- [ ] From https://github.com/flutter/flutter/issues/24760, make sure grouped rows with division gaps, section header/footer labels
- [ ] From https://github.com/flutter/flutter/issues/28052, create a standard tile layout for each row

### PR DESCRIPTION
**Pull request by campovski**
This pull request adds two important widgets that were missing, a `CupertinoListSection` (a close relative of Material Design `ListView`) and a `CupertinoListTile` (a iOS-style twin of Material Design `ListTile`).

D99B6A58-B6E7-4B10-AC3F-CEB00EF5801D
8BDF3105-8FD0-423C-8A94-BA8126057C80
BEB0C1D4-E3D1-49C1-8515-8A3D17A54032
C7A7C150-B6E8-48DC-B1D4-90E142DABAE6

Fixes #19228

Repo [flutter/tests] left intact.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

## Pull Request steps

- [x] Implement `CupertinoListSection` and `CupertinoListTile`.
- [x] Dedup code between these widgets and `CupertinoFormSection` and `CupertinoFormRow`.
- [x] Add tests for `CupertinoListSection` and `CupertinoListTile`. These tests should be similar to the ones already present for `CupertinoFormSection` and `CupertinoFormRow`.
- [ ] Organize PR into meaningful commits.
- [x] Review the code before removing Draft.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 5 untriaged digest(s) on patchset 2.
View them at https://flutter-gold.skia.org/cl/github/78732

**Comment by skia-gold** [comment]
Gold has detected about 5 untriaged digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/78732

**Comment by skia-gold** [comment]
Gold has detected about 5 untriaged digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/78732

**Comment by bsr203** [comment]
hi @campovski do you happen to have the source of demo screens from above available? It looks great and thank you for the work on it.

**Comment by campovski** [comment]
@bsr203 I am glad you are find this useful! I am just sorry that I didn't have so much time lately to finish this PR.

Yes, the examples are available [here](https://github.com/stelynx/flutter-exempli-gratia/tree/master/widgets/cupertino/cupertino_list_section).

**Comment by bsr203** [comment]
@campovski thanks a lot for getting back to me and your time building the examples. I copies your code to latest flutter and it didn't give any error. Will try out the examples of usage. Thanks again for sharing your work. Cheers.

**Comment by skia-gold** [comment]
Gold has detected about 57 new digest(s) on patchset 16.
View them at https://flutter-gold.skia.org/cl/github/78732

File: packages/flutter/lib/src/cupertino/list_tile.dart [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+  final Widget? leading;
+
+  /// A widget displayed at the end of the [CupertinoListTile]. This is usually
+  /// a right chevron icon (e.g. `CupertinoListTileChevron`), or an `Icon`.
+  final Widget? trailing;
+
+  /// The [onTap] function is called when a user taps on [CupertinoListTile]. If
+  /// left `null`, the [CupertinoListTile] will not show any visual information
+  /// when tapped.
+  final Future<void> Function()? onTap;
```
Comment (albertodev01) [inline packages/flutter/lib/src/cupertino/list_tile.dart:221]: **question**: Why is this declared as `final Future<void> Function()? onTap;` rather than `final void Function()? onTap;` (or, in an equivalent form, `final VoidCallback? onTap;`)?

File: packages/flutter/lib/src/cupertino/list_section.dart [hunk 2]
```diff
@@ -0,0 +1,814 @@
...
+///
+/// ```dart
+/// class MyStatefulWidget extends StatefulWidget {
+///   const MyStatefulWidget({Key? key}) : super(key: key);
+///
+///   @override
+///   _MyStatefulWidgetState createState() => _MyStatefulWidgetState();
+/// }
+///
+/// class _MyStatefulWidgetState extends State<MyStatefulWidget> {
```
Comment (albertodev01) [inline packages/flutter/lib/src/cupertino/list_section.dart:145]: **chore**: Since your state class is just overriding `build` and nothing more, make the example with a `StatelessWidget`

File: packages/flutter/lib/src/cupertino/list_section.dart [hunk 3]
```diff
@@ -0,0 +1,814 @@
...
```
Comment (albertodev01) [inline packages/flutter/lib/src/cupertino/list_section.dart:204]: **chore**: Same here, I'd use a `StatelessWidget` instead

File: packages/flutter/lib/src/cupertino/list_section.dart [hunk 4]
```diff
@@ -0,0 +1,814 @@
...
+    double? additionalDividerMargin,
+    this.topMargin = _kMarginTop,
+    bool hasLeading = true,
+  }) : assert((children != null && children.length > 0) || header != null),
+       _type = _CupertinoListSectionType.base,
+       additionalDividerMargin = additionalDividerMargin ??
+           (hasLeading ? _kBaseAdditionalDividerMargin : 0.0),
+       super(key: key);
+
+  /// Creates a section that mimicks standard "Inset Grouped" iOS list section.
```
Comment (albertodev01) [inline packages/flutter/lib/src/cupertino/list_section.dart:322]: **chore**: fix typo

```suggestion
  /// Creates a section that mimics standard "Inset Grouped" iOS list section.
```

**Review by albertodev01 — COMMENTED** [review]
Found some minor things that could be addressed

File: packages/flutter/lib/src/cupertino/list_tile.dart (hunk 1 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/list_tile.dart:221]: You are right. I think I have had an usecase where this needed to be a `Future<void>` but it makes sense to turn this into a `VoidCallback`.

**Comment by Hixie** [comment]
@campovski At a high level this looks great. It probably needs some work e.g. resolving analysis issues and some code review comments and such, but before I get too into the weeds I wanted to ask you what the best way to help you is. Are you still interested in landing this? (I'm painfully aware that it has taken us months to get to it and I'm really sorry about that.)

**Comment by campovski** [comment]
@Hixie Yes, I would like to land this, however unfortunately I haven't got time lately to work on Flutter in general. What I am a bit worried about this PR is organizing it into meaningful commits. I did my best to do so when I actively developed this feature but now it became a bit of a mess with last few commits.

Nonetheless, I promise to resolve the analysis issues and comments tomorrow. Looking forward to it!

**Comment by Hixie** [comment]
Please don't feel like I'm rushing you, I'm obviously in no position to do so since we've been so slow here. Your contributions are massively welcome regardless. :-)

File: packages/flutter/lib/src/cupertino/list_tile.dart (hunk 1 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/list_tile.dart:221]: This should not be a `VoidCallback` for the following reason: pressing a "list tile" on iOS that navigates to a new route, highlights that list tile. Once the route is poped, the list tile is still highlighted in remains to be so, until route is completely popped. If this function completes instantly, given current implementation (simple yet effective) the list tile become not highlighted instantly.

It would make sense to turn this into a `FutureOr<void>` function, however AFAIR this is against Flutter coding standards. I will however still convert this to a `FutureOr<void> Function()? onTap`.

**Comment by campovski** [comment]
@Hixie just wanted to let you know that this PR is now ready.

The analysis errors have been fixed and CR comments either applied or (I believe justiable) answered.

Please, explicitly say, if it is required, that this PR be organised into meaningful commits.

File: packages/flutter/lib/src/cupertino/list_section.dart (hunk 3 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/list_section.dart]: Fixed.

File: packages/flutter/lib/src/cupertino/list_section.dart (hunk 4 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/list_section.dart:311]: Fixed.

File: packages/flutter/lib/src/cupertino/list_section.dart (hunk 2 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/list_section.dart]: Fixed.

**Comment by Hixie** [comment]
No, no need to worry about the commits. It all gets squashed into one when it lands anyway. The only reason to worry about the commits is if doing so would make it notably easier to review (e.g. if one commit is all automated fixes created by a regular expression, or something, vs hand-written code).

**Comment by campovski** [comment]
Ok, that was my thought as well but since this PR affects just a few files, it is probably redundant to organize commits, I agree.

File: packages/flutter/lib/src/cupertino/list_section.dart [hunk 5]
```diff
@@ -0,0 +1,803 @@
...
+///             additionalInfo: const Text('12 days ago'),
+///             trailing: const CupertinoListTileChevron(),
+///             onTap: () => Navigator.of(context).pushNamed('lastCommit'),
+///           ),
+///         ],
+///       ),
+///     );
+///   }
+/// }
+/// ```
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/list_section.dart:235]: for samples, we should probably use the new extracted sample code setup @gspencergoog created
if you look around the code you should find other examples (e.g. AppBar, just search for "See code in examples" in the source).

File: packages/flutter/lib/src/cupertino/list_section.dart [hunk 6]
```diff
@@ -0,0 +1,803 @@
...
+/// If null, defaults to [CupertinoColors.systemGroupedBackground].
+///
+/// {@macro flutter.material.Material.clipBehavior}
+///
+/// See also:
+///
+///  * [CupertinoFormRow], an iOS-style list tile, a typical child of
+///    [CupertinoFormSection].
+///  * [CupertinoListSection], an iOS-style list section.
+class CupertinoFormSection extends StatelessWidget {
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/list_section.dart:619]: Just out of interest, why did you move this out of its own file? Are there private fields that need sharing or something?

File: packages/flutter/lib/src/cupertino/list_section.dart (hunk 5 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/cupertino/list_section.dart:235]: There's also a [README](https://github.com/flutter/flutter/blob/master/examples/api/README.md#authoring) for this.

**Comment by christopherfujino** [comment]
@campovski are you still interested in landing this? if so, you will need to address code review feedback.

File: packages/flutter/lib/src/cupertino/list_section.dart (hunk 6 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/list_section.dart:619]: Yes, I agree. I will wager different options today and get back to you in the evening.

**Comment by campovski** [comment]
@christopherfujino Yes, I am, I just somehow missed the comments from Hixie.

File: packages/flutter/lib/src/cupertino/list_section.dart (hunk 5 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/list_section.dart:235]: I think this [Snippets](https://github.com/flutter/flutter/blob/master/examples/api/README.md#snippets) is a better one for this case.

Comment (campovski) [reply packages/flutter/lib/src/cupertino/list_section.dart:234]: This is the [commit](https://github.com/flutter/flutter/pull/78732/commits/c6c2225522f83b43dfd8da2a59aad249e9132364) I made regarding this issue.

File: packages/flutter/lib/src/cupertino/list_section.dart (hunk 6 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/list_section.dart]: Is [this](https://github.com/flutter/flutter/pull/78732/commits/f521cd31fd36c3174b7cd5537386800a2198b751) going to be OK?

**Comment by vnagendra** [comment]
I was looking for a nice `CupertinoListTile`. This seemed like a good one to experiment with. @campovski  - In iOS when you have a `title` (or `subtitle`) that has a long text -- aka longer than the width allows, you have a couple of options. Either to wrap or truncate. 

Material's `ListTile` seems to wrap it, perfectly sensible. This new tile doesn't wrap or truncate. It overflows. 

I didn't want to figure out how to provide a test case for a PR. But if you change any of your list tiles to have the standard Flutter example, you'll see it too.. In my case I had the following wrapped in a `ListView` inside a `CupertinoPageScaffold`. 

```
        children: [
          const CupertinoListTile(
            title: Text("You have pushed the button this many times."),
          ),
          CupertinoListTile(
              title: Text(
            "$_counter",
            style: Theme.of(context).textTheme.headline4,
          )),
```

Error I saw was 
```
======== Exception caught by rendering library =====================================================
The following assertion was thrown during layout:
A RenderFlex overflowed by 2.0 pixels on the right.
```

**Comment by campovski** [comment]
@vnagendra Yes, this is intentional. I am pretty sure that is also mentioned in a docstring, but I might be wrong. If it is not, I will have to add it.

File: packages/flutter/lib/src/cupertino/form_section.dart [hunk 7]
```diff
@@ -65,8 +54,14 @@ enum _CupertinoFormSectionType { base, insetGrouped }
...
 /// {@macro flutter.material.Material.clipBehavior}
+///
+/// See also:
+///
+///  * [CupertinoFormRow], an iOS-style list tile, a typical child of
+///    [CupertinoFormSection].
+///  * [CupertinoListSection], an iOS-style list section.
 class CupertinoFormSection extends StatelessWidget {
-  /// Creates a section that mimics standard iOS forms.
+  /// Creates a section that mimicks standard iOS forms.
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/form_section.dart:64]: looks like "mimics" is actually the more normal spelling (though mimicks is right too)

File: packages/flutter/lib/src/cupertino/form_section.dart [hunk 8]
```diff
@@ -107,11 +102,11 @@ class CupertinoFormSection extends StatelessWidget {
     this.backgroundColor = CupertinoColors.systemGroupedBackground,
     this.decoration,
     this.clipBehavior = Clip.none,
-  })  : _type = _CupertinoFormSectionType.base,
+  })  : _type = CupertinoListSectionType.base,
         assert(children.length > 0),
         super(key: key);
 
-  /// Creates a section that mimics standard "Inset Grouped" iOS forms.
+  /// Creates a section that mimicks standard "Inset Grouped" iOS forms.
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/form_section.dart:109]: (as above)

File: packages/flutter/lib/src/cupertino/list_tile.dart [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+
+    if (widget.onTap == null)
+      return child;
+
+    return GestureDetector(
+      onTapDown: (_) => setState(() => _tapped = true),
+      onTapCancel: () => setState(() => _tapped = false),
+      onTap: () async {
+        await widget.onTap!();
+        setState(() => _tapped = false);
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/list_tile.dart:389]: this is a trivial nit but generally we prefer `setState(() { ... })` since it doesn't actually return anything

**Comment by Hixie** [comment]
This is great, and sorry for how long it took to review.

**Comment by Hixie** [comment]
@campovski Only some trivial nits (and branch conflicts); once you've had a chance to fix those please poke me (either on Discord or e-mail me — I get too many GitHub notifications to notice comments on GitHub unfortunately) and I'll make sure it lands. Thanks!

File: packages/flutter/lib/src/cupertino/list_tile.dart (hunk 9 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/list_tile.dart:389]: Will change all such occurrences.

File: packages/flutter/lib/src/cupertino/form_section.dart (hunk 7 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/form_section.dart:64]: Ok, I will change back to mimics.

File: packages/flutter/lib/src/cupertino/form_section.dart (hunk 8 above)
Comment (campovski) [reply packages/flutter/lib/src/cupertino/form_section.dart:109]: same

**Comment by alex-sandri** [comment]
@campovski I know this behavior is intentional, but is there any reason why wrapping the `Column` that contains both the title and the subtitle in https://github.com/flutter/flutter/blob/451ccced68e4d7f45204e516271e6f349a6ba381/packages/flutter/lib/src/cupertino/list_tile.dart#L356 with an `Expanded` widget and then removing the `Spacer` in https://github.com/flutter/flutter/blob/451ccced68e4d7f45204e516271e6f349a6ba381/packages/flutter/lib/src/cupertino/list_tile.dart#L367 would not be a better (correct me if I'm wrong) solution?

This would solve overflow issues like https://github.com/flutter/flutter/pull/78732#issuecomment-1093133652 and as far as I can tell wouldn't result in any direct downsides as restricting the title / subtitle extent would still be possible. IMHO this should be, if not the default behavior, at least provided as an option via a parameter for example.

**Comment by campovski** [comment]
@alex-sandri @henry2man I agree, we should change it.

**Review by Hixie — APPROVED** [review]
LGTM

Thank you so much for this contribution!

**Comment by Hixie** [comment]
(sorry for the delay, i need to get a second reviewer to check it over per our process, and a lot of people are on vacation)

File: examples/api/lib/cupertino/list_section/list_section_base.0.dart [hunk 10]
```diff
@@ -0,0 +1,95 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (Piinks) [inline examples/api/lib/cupertino/list_section/list_section_base.0.dart:1]: Can you add tests for your samples? They help us ensure we don't have broken sample code in our documentation.

You can find examples of these tests in this directory, which is where you'll want to put your tests as well: https://github.com/flutter/flutter/tree/451ccced68e4d7f45204e516271e6f349a6ba381/examples/api/test

Comment (campovski) [reply examples/api/lib/cupertino/list_section/list_section_base.0.dart:1]: But of course, will try to add them tomorrow morning.

Comment (Piinks) [reply examples/api/lib/cupertino/list_section/list_section_base.0.dart:1]: Thank you! 🙏

Comment (campovski) [reply examples/api/lib/cupertino/list_section/list_section_base.0.dart:1]: Unfortunately, I was not able to add tests this morning so I will have to postpone this till evening due to urgent errands.

Comment (campovski) [reply examples/api/lib/cupertino/list_section/list_section_base.0.dart:1]: I have added the tests - hopefully this is what you were looking for.

Comment (Piinks) [reply examples/api/lib/cupertino/list_section/list_section_base.0.dart:1]: Awesome! Thank you!

**Review by Piinks — APPROVED** [review]
Flutter_LGTM

Thank you!

**Comment by Piinks** [comment]
The google testing shard here is stuck, I have confirmed the tests passed. Merging manually.

**Comment by manuel-plavsic** [comment]
Which flutter stable version will include these new widgets?

**Comment by jmagman** [comment]
This merged in commit https://github.com/flutter/flutter/commit/0c40945a67dc936f7eb61b7f752d2cec5fbe4055 and right now is only available in the master channel.  You can follow along on that commit as tags get added to it https://github.com/flutter/flutter/wiki/Where's-my-Commit%3F#finding-the-versions-that-contain-framework-commit-x
